### ISSUE TITLE
invalid pattern strings are not flagged

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/rosetta#59]
Related to https://github.com/phetsims/rosetta/issues/44.

See screenshot below from translation of "Concentration".  The English string is "{0} {1}".  I entered "{0}", which is clearly incomplete, then tabbed to the next field.  I got no feedback that the translated string was invalid. 

![screenshot_592](https://cloud.githubusercontent.com/assets/3046552/9613863/429e0f14-50ab-11e5-8bd2-8925d556af7e.png)

In the Java version, the user gets immediate feedback. Example screenshot below. When the text field loses focus, the value is validated.  If it's invalid, it's flagged in red. Clicking on the red "error" icon opens a dialog explaining the problem.

![screenshot_593](https://cloud.githubusercontent.com/assets/3046552/9613742/8c3a701e-50aa-11e5-9585-7d7c5bef0606.png)

**Comment by pixelzoom** [comment]
The Java version provides similar feedback for string containing HTML markup, identifying missing/unmatched tags.

**Comment by aaronsamuel137** [comment]
I've added a similar behavior as the Java sims. Assigning to @pixelzoom for review.

**Comment by pixelzoom** [comment]
Looks great, but I think we should discuss the error message and terminology, which differs from the Java version.

Internally, we refer to these as "MessageFormat placeholders". The error message in the Java version identifies specifically which placeholder(s) are missing. For example, if the English string is "{0} {1} {2}" and the translated string is "{0}", then the error message is:

`missing MessageFormatPlaceholders: {1} {2}`

Error message in the HTML5 version is:

`You must submit the same pattern replacements as the English string`

Questions:

(1) Is there a good reason to use different terminology than the Java version?

(2) How is this described in documentation? Will translators know what a "pattern replacement" is?  

(3) Is it useful to indicate which placeholders are missing, or is a generic message OK?

**Comment by aaronsamuel137** [comment]
It's not difficult to use the same message as the Java, so I've changed it.

Regarding point 2, I think we still might want to document what "MessageFormat placeholders" are. Perhaps @phet-steele can do this when he creates the document.

Assigning to @pixelzoom for review.

**Comment by pixelzoom** [comment]
Looks good for missing placeholders. How about extra placeholders?  Eg, English is "{0}" and translated is "{0} {1}"?  Do we need to flag that?

**Comment by phet-steele** [comment]
@pixelzoom @aaronsamuel137 I just put a note of explaining MessageFormat placeholders in the document.

**Comment by aaronsamuel137** [comment]
Added validation for extra placeholders too. A few caveats:

1. It won't handle 10 or more MessageFormat placeholders, due to the way that regex is structured. I assume this will never be an issues?
2. What about the case of two of the same number, i.e. {0}{0}? Should this be allowed? Currently it is.

Assigning to @pixelzoom for review.

**Comment by pixelzoom** [comment]
This doesn't seem to be correctly detecting missing or extra placeholders. See screenshot below for Arabic (ar) translation of Concentration.  Is the fact that Arabic is RTL causing problems?

![screenshot_608](https://cloud.githubusercontent.com/assets/3046552/9770110/63a024f4-56eb-11e5-8713-b3f4111f3b4b.png)

About your questions:

> 1. It won't handle 10 or more MessageFormat placeholders, due to the way that regex is structured. I assume this will never be an issues?

I don't understand why there should be a limit. This sounds like a problem with your regex. What is the regex?

> 2. What about the case of two of the same number, i.e. {0}{0}? Should this be allowed? Currently it is.

It shouldn't be allowed. Each placeholder that occurs in the English string should appear exactly once in the translated string. And the translated string should contain no placeholders that don't appear in the English string.

**Comment by aaronsamuel137** [comment]
Arabic concentration was working for me, but I did discover a bug where the error message wouldn't get updated if you had one error, and then tabbed back in and created a different error. Maybe that's what was happening for you? Either way, I think it is fixed now and didn't have to do with the rtl.

I've also addressed the other points. Duplicates will now show up as "extra" messageFormats, just like ones that shouldn't be there to begin with.

Currently, you do need to tab away from the field for the error message to be updated if you are going from one error to another.

Assigning to @pixelzoom for review.
